### ISSUE TITLE
[BUG] salt-ssh ModuleNotFoundError: No module named 'jinja2.asyncsupport'

### ISSUE THREAD

**Issue by daro1337** [saltstack/salt#60035]
**Description**
`salt-ssh 'srv1' state.apply check_pl --python3-bin=/usr/bin/python`
returns:

`Traceback (most recent call last):
  File "/var/tmp/.root_040010_salt/salt-call", line 27, in <module>
    salt_call()
  File "/var/tmp/.root_040010_salt/pyall/salt/scripts.py", line 437, in salt_call
    import salt.cli.call
  File "/var/tmp/.root_040010_salt/pyall/salt/cli/call.py", line 6, in <module>
    import salt.cli.caller
  File "/var/tmp/.root_040010_salt/pyall/salt/cli/caller.py", line 15, in <module>
    import salt.minion
  File "/var/tmp/.root_040010_salt/pyall/salt/minion.py", line 25, in <module>
    import salt.client
  File "/var/tmp/.root_040010_salt/pyall/salt/client/__init__.py", line 47, in <module>
    import salt.utils.minions
  File "/var/tmp/.root_040010_salt/pyall/salt/utils/minions.py", line 12, in <module>
    import salt.auth.ldap
  File "/var/tmp/.root_040010_salt/pyall/salt/auth/ldap.py", line 18, in <module>
    from jinja2 import Environment
  File "/var/tmp/.root_040010_salt/pyall/jinja2/__init__.py", line 81, in <module>
    _patch_async()
  File "/var/tmp/.root_040010_salt/pyall/jinja2/__init__.py", line 77, in _patch_async
    from jinja2.asyncsupport import patch_all
ModuleNotFoundError: No module named 'jinja2.asyncsupport'`

**Setup**
salt-ssh from saltmaster (3002.5 @ debian9)
remote: ubuntu 18.04LTS

**Steps to Reproduce the behavior**

run state.apply or cmd.run via salt-ssh

**Expected behavior**

cmd.run /state.apply succeeed

**Screenshots**

**Versions Report**
```
Salt Version:
          Salt: 3002.5
 
Dependency Versions:
          cffi: 1.13.2
      cherrypy: 3.5.0
      dateutil: 2.5.3
     docker-py: 1.9.0
         gitdb: 2.0.0
     gitpython: 2.1.1
        Jinja2: 2.9.4
       libgit2: 0.28.4
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.19
      pycrypto: 2.6.1
  pycryptodome: 3.6.1
        pygit2: 1.0.2
        Python: 3.5.3 (default, Nov 18 2020, 21:09:16)
  python-gnupg: Not Installed
        PyYAML: 3.12
         PyZMQ: 17.1.2
         smmap: 2.0.1
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.2.1
 
System Versions:
          dist: debian 9 stretch
        locale: UTF-8
       machine: x86_64
       release: 4.9.0-14-amd64
        system: Linux
       version: Debian GNU/Linux 9 stretch

```
</details>

**Additional context**

It seems that salt-ssh does not send all modules files from jinja2 library. Because there is a jinja2 lib in /var/tmp/.root_040010_salt/pyall but without asyncsupport.py it throws this error error. Even if I have python3-jinja2 installed on remote server, it does not search jinja2 lib  in another python path e.g /usr/lib/python3/dist-packages/

`/usr/lib/python3/dist-packages/jinja2/asyncsupport.py`

PoC:
`root@srv1:/var/tmp/.root_040010_salt# python3
Python 3.6.9 (default, Jan 26 2021, 15:33:00) 
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.path
['', '/usr/lib/python36.zip', '/usr/lib/python3.6', '/usr/lib/python3.6/lib-dynload', '/usr/local/lib/python3.6/dist-packages', '/usr/lib/python3/dist-packages']
>>> from jinja2.asyncsupport import patch_all
>>> 

`

`
root@srv1:/var/tmp/.root_040010_salt# python3
Python 3.6.9 (default, Jan 26 2021, 15:33:00) 
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.path.insert(0, '/var/tmp/.root_040010_salt/pyall')
>>> sys.path
['/var/tmp/.root_040010_salt/pyall', '', '/usr/lib/python36.zip', '/usr/lib/python3.6', '/usr/lib/python3.6/lib-dynload', '/usr/local/lib/python3.6/dist-packages', '/usr/lib/python3/dist-packages']
>>> 
>>> from jinja2.asyncsupport import patch_all
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/var/tmp/.root_040010_salt/pyall/jinja2/__init__.py", line 81, in <module>
    _patch_async()
  File "/var/tmp/.root_040010_salt/pyall/jinja2/__init__.py", line 77, in _patch_async
    from jinja2.asyncsupport import patch_all
ModuleNotFoundError: No module named 'jinja2.asyncsupport'`

**Comment by daro1337** [comment]
temporary workaround on remote server:
```
rm /var/tmp/.root_040010_salt/pyall/jinja2
ln -s /usr/lib/python3/dist-packages/jinja2 /var/tmp/.root_040010_salt/pyall/jinja2
```

**Comment by kylehuff** [comment]
We were able to solve this problem simply by installing python3-jinja2 version 2.10 on the salt-master. Nothing needed to be changed on the ssh-minion, or in salt-ssh.

The debian 9 (stretch) repository only has python3-jinja2 v2.8, but the saltstack debian 9 APT repository includes the upstream version 2.9.4 that includes the async patch (the patched version should have a method called '_patch_async' in /usr/lib/python3/dist-packages/jinja2/__init__.py)

If you are not using the staltstack repo, Jinaj2 v2.10 is available via pip3:

`pip3 install jinja2==2.10`

Here is the relevant version output:

```
pip3 freeze | grep Jinja
Jinaj2==2.10
```

```
Dependency Versions:
...
        Jinja2: 2.10
...
 
System Versions:
      dist: debian 9.13
```

Ultimately, I believe the problem lies with python3-jinja2, not the salt master, salt-ssh or the remote host - so the real fix is to have the salt-master use the corrected version of python3-jinja2, and whatever is deployed to the remote system should (IMO) remain irrelevant.

One of the things that makes salt-ssh so useful is that is uses as little of the python framework on the remote client as possible -- if salt-ssh starts pulling in a bunch of libraries from the remote I see that resulting in unpredictable behavior.

So my question is, what does "Approved" in this process mean? Are changes to how salt-ssh loads modules on the remote host being planned/developed? If so, what has been proposed? I'm not saying it can't or even that it shouldn't be done, just saying that in this example, the issue is a broken Jinja2 module -- not a broken saltstack subsystem (unless I am missing something).

**Comment by daro1337** [comment]
@kylehuff I'm using saltstack repo. Unfortunately, in saltstack repo for Debian 9 and salt-master version 3002.x as you noticed there is 2.9.4 version of python3-jinja2, That version is installed on my saltmaster with cause this error.
```
python3-jinja2 2.9.4-1 
```
Updating jinja2 to 2.10 via pip3 resolved the issue

**Comment by kylehuff** [comment]
@daro1337 You are correct, the salt-stack repo version of jinja2 is not patched for this issue, I must have manually patched it at some point. I updated my comment to not confuse anyone who ends up here looking for the fix.

**Comment by daro1337** [comment]
@kylehuff  should I mark this Issue as resolved since workaround exist (upgrade jinja2 via pip) or repo for debian should be updated?
